File: sample-scala/app/controllers/Application.scala
package controllers

import java.io.ByteArrayOutputStream
import java.nio.file.Files
import java.nio.file.Paths

import scala.collection.JavaConverters.collectionAsScalaIterableConverter
import scala.collection.immutable.ListMap
import scala.util.matching.Regex

import org.apache.fop.apps.FOUserAgent
import org.apache.fop.apps.Fop
import org.apache.fop.fo.FOTreeBuilder
import org.apache.fop.fonts.FontInfo
import org.apache.fop.fonts.Typeface
import org.apache.xmlgraphics.util.MimeConstants

import com.dmanchester.playfop.api.Units
import com.dmanchester.playfop.sapi.PlayFop

import javax.inject.Inject
import models.Label
import models.PaperSizeAndWhiteSpace
import play.api.Configuration
import play.api.data.Form
import play.api.data.Forms.mapping
import play.api.data.Forms.number
import play.api.data.Forms.optional
import play.api.data.Forms.text
import play.api.i18n.I18nSupport
import play.api.mvc.AbstractController
import play.api.mvc.ControllerComponents
import views.util.Calc

class Application @Inject() (config: Configuration, cc: ControllerComponents, val playFop: PlayFop)
    extends AbstractController(cc) with I18nSupport {

  private val AboutPageAddlInfoProperty = "about.page.addl.info"
  private val FontFamilyExclusionRegexProperty = "font.family.exclusion.regex"
  private val InitialFontFamilyProperty = "initial.font.family"

  private val SheetSizeAndWhiteSpaceInMM = new PaperSizeAndWhiteSpace(
      height = 297 /* A4 */, width = 210 /* A4 */, margin = 20,
      interLabelGutter = 10, intraLabelPadding = 2)

  private val mm = new Units("mm", 1)
  private val SheetRows = 10
  private val SheetCols = 3
  private val SheetFilename = "labels-sheet.pdf"
  private val SheetPdfCreator = "PlayFOP Labels, Scala version"  // producer already references Apache FOP, so no need to do so again here

  private val FontSizesInPoints = (6 to 18)
  private val ImageNameCityscape = "Cityscape"
  private val ImageNameSunrise = "Sunrise"
  private val ImageNameVolcano = "Volcano"
  private val ImageNamesToPaths = ListMap(
    ImageNameCityscape -> "images/emoji_u1f306_cityscape.svg",
    ImageNameSunrise -> "images/emoji_u1f304_sunrise.svg",
    ImageNameVolcano -> "images/emoji_u1f30b_volcano.svg"
  )

  private val InitialText =
    """John Smith
      |123 Main St.
      |Anytown, MA  09876""".stripMargin
  private val InitialFontSizeInPoints = 9
  private val InitialImageName = ImageNameCityscape

  private val SingleLabelScaleFactor = 3

  private val PlayFopUrl = "https://www.dmanchester.com/playfop"

  private val LabelForm: Form[Label] = Form(
    mapping(
      "text" -> text(maxLength = Label.TextMaxLength),
      "fontFamily" -> text,
      "fontSizeInPoints" -> number,
      "imageName" -> optional(text)
    )(Label.apply)(Label.unapply)
  )

  private val fontFamilies = getFontFamilies()  // getFontFamilies() is an expensive operation, so cache its result

  private def getFontFamilies(): List[String] = {

    val fop: Fop = playFop.newFop(MimeConstants.MIME_PDF, new ByteArrayOutputStream(), autoDetectFontsForPDF = true)

    val fontInfo: FontInfo = fop.getDefaultHandler().asInstanceOf[FOTreeBuilder].getEventHandler().getFontInfo()

    val typefaces: Iterable[Typeface] = fontInfo.getFonts().values().asScala

    val fontNamesUnfiltered = typefaces.map(_.getFullName()).toList

    val fontFamilyExclusionRegex: Option[Regex] = config.get[Option[String]](FontFamilyExclusionRegexProperty).map(_.r)

    // If an exclusion regex was provided, filter out any font names that
    // match it.
    val fontNames = fontFamilyExclusionRegex.map { regex =>
      fontNamesUnfiltered.filterNot { fontName =>
        regex.findFirstIn(fontName).isDefined
      }
    }.getOrElse(fontNamesUnfiltered)

    fontNames.sorted
  }

  def index() = Action {
    Redirect(routes.Application.designLabels())
  }

  def designLabels() = Action { implicit request =>
    val imageNames = "" :: ImageNamesToPaths.keys.toList  // at top of list, offer a no-image option
    Ok(views.html.labelDesign(getInitialForm(), fontFamilies, getFontSizesAndText(), imageNames))
  }

  private def getInitialForm(): Form[Label] = {

    val initialFontFamily: Option[String] = config.get[Option[String]](InitialFontFamilyProperty)

    // If the property has been set, confirm that the font family is available.
    initialFontFamily.foreach { initialFontFamily =>
      if (!fontFamilies.contains(initialFontFamily)) {
        throw new IllegalArgumentException(s"Font family $initialFontFamily not found!")
      }
    }

    val label =  Label(
      InitialText,
      initialFontFamily.getOrElse(fontFamilies.head),  // if property not set, use first font family in list
      InitialFontSizeInPoints,
      Some(InitialImageName)
    )

    LabelForm.fill(label)
  }

  private def getFontSizesAndText(): Map[String, String] = {
    val fontSizesAndText = FontSizesInPoints.map { size => (size.toString, size + " points") }
    ListMap(fontSizesAndText:_*)
  }

  def generateSingleLabelAsPNG() = Action { implicit request =>

    LabelForm.bindFromRequest.fold(
      formWithErrors => {
        // UI prevents user from entering bad input; only a hand-rolled input
        // URL can trigger errors. We want to alert user to those errors, but we
        // don't bother with a nice presentation.
        BadRequest(formWithErrors.errorsAsJson)
      },
      label => {

        val imageURI: Option[String] = getImageURI(label.imageName)

        val mimeType = MimeConstants.MIME_PNG

        val labelWidthInMM = SingleLabelScaleFactor * Calc.labelWidth(SheetSizeAndWhiteSpaceInMM, SheetCols)
        val labelHeightInMM = SingleLabelScaleFactor * Calc.labelHeight(SheetSizeAndWhiteSpaceInMM, SheetRows)
        val intraLabelPaddingInMM = SingleLabelScaleFactor * SheetSizeAndWhiteSpaceInMM.intraLabelPadding

        Ok(
          playFop.processTwirlXml(
            views.xml.labelSingle.render(labelWidthInMM, labelHeightInMM, intraLabelPaddingInMM, mm, imageURI, label.scale(SingleLabelScaleFactor)),
            mimeType
          )
        ).as(mimeType)
      }
    )
  }

  private def getImageURI(imageName: Option[String]): Option[String] = {

    val imagePath: Option[String] = imageName.flatMap(ImageNamesToPaths.get(_))

    imagePath.map(thePath => this.getClass().getClassLoader().getResource(thePath).toString())
  }

  def generateLabelsSheetAsPDF() = Action { implicit request =>

    LabelForm.bindFromRequest.fold(
      formWithErrors => {
        // See comment above about not bothering with a nice presentation.
        BadRequest(formWithErrors.errorsAsJson)
      },
      label => {

        val imageURI: Option[String] = getImageURI(label.imageName)

        val mimeType = MimeConstants.MIME_PDF

        val foUserAgentBlock = { foUserAgent: FOUserAgent =>
          foUserAgent.setCreator(SheetPdfCreator)
        }

        Ok(
          playFop.processTwirlXml(
            views.xml.labelsSheet.render(SheetSizeAndWhiteSpaceInMM, mm, SheetRows, SheetCols, imageURI, label),
            mimeType,
            autoDetectFontsForPDF = true,
            foUserAgentBlock = foUserAgentBlock
          )
        ).as(mimeType).withHeaders(
          CONTENT_DISPOSITION -> s"attachment; filename=$SheetFilename"
        )
      }
    )
  }

  def showAbout() = Action {

    val addlInfoPath: Option[String] = config.get[Option[String]](AboutPageAddlInfoProperty)

    val addlInfoAsHtml: Option[String] = addlInfoPath.map { thePath =>

      val addlInfoPathObj = Paths.get(thePath)
      val addlInfoAsBytes = Files.readAllBytes(addlInfoPathObj)
      new String(addlInfoAsBytes, "utf-8")
    }

    Ok(views.html.about(PlayFopUrl, addlInfoAsHtml))
  }
}

File: sample-java/app/controllers/Application.java
package controllers;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.apache.fop.apps.FOPException;
import org.apache.fop.apps.FOUserAgent;
import org.apache.fop.apps.Fop;
import org.apache.fop.fo.FOTreeBuilder;
import org.apache.fop.fonts.FontInfo;
import org.apache.fop.fonts.Typeface;
import org.apache.xmlgraphics.util.MimeConstants;

import com.dmanchester.playfop.api.Units;
import com.dmanchester.playfop.japi.FOUserAgentBlock;
import com.dmanchester.playfop.japi.PlayFop;
import com.dmanchester.playfop.japi.ProcessOptions;
import com.typesafe.config.Config;

import models.Label;
import models.PaperSizeAndWhiteSpace;
import play.data.Form;
import play.data.FormFactory;
import play.mvc.Controller;
import play.mvc.Http.HeaderNames;
import play.mvc.Result;
import views.util.Calc;

public class Application extends Controller {

    private static final String ABOUT_PAGE_ADDL_INFO_CONFIG_PATH = "about.page.addl.info";
    private static final String FONT_FAMILY_EXCLUSION_REGEX_CONFIG_PATH = "font.family.exclusion.regex";
    private static final String INITIAL_FONT_FAMILY_CONFIG_PATH = "initial.font.family";

    private static final PaperSizeAndWhiteSpace SHEET_SIZE_AND_WHITESPACE_IN_MM =
            new PaperSizeAndWhiteSpace(297 /* height (A4) */,
            210 /* width (A4) */, 20 /* margin */, 10 /* interLabelGutter */,
            2 /* intraLabelPadding */);

    private static final Units MM = new Units("mm", 1);
    private static final int SHEET_ROWS = 10;
    private static final int SHEET_COLS = 3;
    private static final String SHEET_FILENAME = "labels-sheet.pdf";
    private static final String SHEET_PDF_CREATOR = "PlayFOP Labels, Java version";  // producer already references Apache FOP, so no need to do so again here

    private static final int FONT_SIZE_IN_POINTS__START = 6;
    private static final int FONT_SIZE_IN_POINTS__END = 18;
    private static final String IMAGE_NAME__CITYSCAPE = "Cityscape";
    private static final String IMAGE_NAME__SUNRISE = "Sunrise";
    private static final String IMAGE_NAME__VOLCANO = "Volcano";
    private static final Map<String, String> IMAGE_NAMES_TO_PATHS = Collections.unmodifiableMap(getImageNamesToPaths());

    private static final String INITIAL_TEXT =
            "John Smith\n" +
            "123 Main St.\n" +
            "Anytown, MA  09876";
    private static final Integer INITIAL_FONT_SIZE_IN_POINTS = 9;
    private static final String INITIAL_IMAGE_NAME = IMAGE_NAME__CITYSCAPE;

    private static final int SINGLE_LABEL_SCALE_FACTOR = 3;

    private static final String PLAYFOP_URL = "https://www.dmanchester.com/playfop";

    private static Map<String, String> getImageNamesToPaths() {

        Map<String, String> imageNamesToPaths = new LinkedHashMap<String, String>();

        imageNamesToPaths.put(IMAGE_NAME__CITYSCAPE, "images/emoji_u1f306_cityscape.svg");
        imageNamesToPaths.put(IMAGE_NAME__SUNRISE, "images/emoji_u1f304_sunrise.svg");
        imageNamesToPaths.put(IMAGE_NAME__VOLCANO, "images/emoji_u1f30b_volcano.svg");

        return imageNamesToPaths;
    }

    private Config config;
    private FormFactory formFactory;
    private PlayFop playFop;
    private List<String> fontFamilies;

    public Application(Config config, FormFactory formFactory, PlayFop playFop) {

        this.config = config;
        this.formFactory = formFactory;
        this.playFop = playFop;

        this.fontFamilies = getFontFamilies();  // getFontFamilies() is an expensive operation, so cache its result
    }

    private List<String> getFontFamilies() {

        ProcessOptions processOptions = new ProcessOptions.Builder().
                autoDetectFontsForPDF(true).build();

        Fop fop = playFop.newFop(MimeConstants.MIME_PDF, new ByteArrayOutputStream(), processOptions);

        FontInfo fontInfo;
        try {
            fontInfo = ((FOTreeBuilder)fop.getDefaultHandler()).getEventHandler().getFontInfo();
        } catch (FOPException ex) {
            throw new RuntimeException(ex);
        }

        Collection<Typeface> typefaces = fontInfo.getFonts().values();

        List<String> fontNamesUnfiltered = new ArrayList<String>(typefaces.size());
        for (Typeface typeface : typefaces) {
            fontNamesUnfiltered.add(typeface.getFullName());
        }

        List<String> fontNames;

        // If an exclusion regex was provided, filter out any font names that
        // match it.
        if (config.hasPath(FONT_FAMILY_EXCLUSION_REGEX_CONFIG_PATH)) {

            fontNames = new ArrayList<String>();

            String fontFamilyExclusionRegex = config.getString(FONT_FAMILY_EXCLUSION_REGEX_CONFIG_PATH);
            Pattern pattern = Pattern.compile(fontFamilyExclusionRegex);

            for (String fontName : fontNamesUnfiltered) {
                Matcher matcher = pattern.matcher(fontName);
                if (!matcher.find()) {
                    fontNames.add(fontName);
                }
            }

        } else {

            fontNames = fontNamesUnfiltered;

        }

        Collections.sort(fontNames);

        return fontNames;
    }

    public Result index() {
        return redirect(controllers.routes.Application.designLabels());
    }

    public Result designLabels() {

        List<String> imageNames = new ArrayList<String>(IMAGE_NAMES_TO_PATHS.keySet());
        imageNames.add(0, "");  // offer a no-image option

        return ok(views.html.labelDesign.render(getInitialForm(), fontFamilies, getFontSizesAndText(), imageNames));
    }

    private Form<Label> getInitialForm() {

        Label initialLabel = new Label();
        initialLabel.setText(INITIAL_TEXT);
        initialLabel.setFontSizeInPoints(INITIAL_FONT_SIZE_IN_POINTS);
        initialLabel.setImageName(INITIAL_IMAGE_NAME);

        if (config.hasPath(INITIAL_FONT_FAMILY_CONFIG_PATH)) {

            // The property has been set. Get its value and confirm that the
            // font family is available.

            String initialFontFamily = config.getString(INITIAL_FONT_FAMILY_CONFIG_PATH);

            if (fontFamilies.contains(initialFontFamily)) {

                initialLabel.setFontFamily(initialFontFamily);

            } else {

                String message = String.format("Font family %s not found!", initialFontFamily);
                throw new IllegalArgumentException(message);
            }

        } else {

            // The property has not been set. Use the first font family in the
            // list.
            initialLabel.setFontFamily(fontFamilies.get(0));
        }

        return formFactory.form(Label.class).fill(initialLabel);
    }

    private Map<String, String> getFontSizesAndText() {

        Map<String, String> fontSizes = new LinkedHashMap<String, String>(FONT_SIZE_IN_POINTS__END - FONT_SIZE_IN_POINTS__START + 1);

        for (int i = FONT_SIZE_IN_POINTS__START; i <= FONT_SIZE_IN_POINTS__END; i++) {
            fontSizes.put(Integer.toString(i), i + " points");
        }

        return fontSizes;
    }

    public Result generateSingleLabelAsPNG() {

        Form<Label> labelForm = formFactory.form(Label.class).bindFromRequest();
        if (labelForm.hasErrors()) {
            // UI prevents user from entering bad input; only a hand-rolled
            // input URL can trigger errors. We want to alert user to those
            // errors, but we don't bother with a nice presentation.
            return badRequest(labelForm.errorsAsJson());
        }

        Label label = labelForm.get();

        String imageURI = getImageURI(label.getImageName());
        // In the case of a null imageURI, template simply doesn't show an
        // image.

        String mimeType = MimeConstants.MIME_PNG;

        double labelWidthInMM = SINGLE_LABEL_SCALE_FACTOR * Calc.getLabelWidth(SHEET_SIZE_AND_WHITESPACE_IN_MM, SHEET_COLS);
        double labelHeightInMM = SINGLE_LABEL_SCALE_FACTOR * Calc.getLabelHeight(SHEET_SIZE_AND_WHITESPACE_IN_MM, SHEET_ROWS);
        double intraLabelPaddingInMM = SINGLE_LABEL_SCALE_FACTOR * SHEET_SIZE_AND_WHITESPACE_IN_MM.getIntraLabelPadding();

        byte[] pngBytes = playFop.processTwirlXml(
            views.xml.labelSingle.render(labelWidthInMM, labelHeightInMM, intraLabelPaddingInMM, MM, imageURI, label.scale(SINGLE_LABEL_SCALE_FACTOR)),
            mimeType
        );

        return ok(pngBytes).as(mimeType);
    }

    private String getImageURI(String imageName) {

        String imagePath = IMAGE_NAMES_TO_PATHS.get(imageName);
        String imageURI = (imagePath == null) ? null :
            Application.class.getClassLoader().getResource(imagePath).toString();

        return imageURI;
    }

    public Result generateLabelsSheetAsPDF() {

        Form<Label> labelForm = formFactory.form(Label.class).bindFromRequest();
        if (labelForm.hasErrors()) {
            // See comment above about not bothering with a nice presentation.
            return badRequest(labelForm.errorsAsJson());
        }

        Label label = labelForm.get();

        String imageURI = getImageURI(label.getImageName());
        // See comment above about a null imageURI leading the template to not
        // show an image.

        String mimeType = MimeConstants.MIME_PDF;

        FOUserAgentBlock foUserAgentBlock = new FOUserAgentBlock() {
            @Override
            public void withFOUserAgent(FOUserAgent foUserAgent) {
                foUserAgent.setCreator(SHEET_PDF_CREATOR);
            }
        };

        ProcessOptions processOptions = new ProcessOptions.Builder().
                autoDetectFontsForPDF(true).foUserAgentBlock(foUserAgentBlock).build();

        String contentDispHeader = String.format("attachment; filename=%s", SHEET_FILENAME);
        response().setHeader(HeaderNames.CONTENT_DISPOSITION, contentDispHeader);

        byte[] pdfBytes = playFop.processTwirlXml(
            views.xml.labelsSheet.render(SHEET_SIZE_AND_WHITESPACE_IN_MM, MM, SHEET_ROWS, SHEET_COLS, imageURI, label),
            mimeType,
            processOptions
        );

        return ok(pdfBytes).as(mimeType);
    }

    public Result showAbout() {

        String addlInfoAsHtml;

        if (config.hasPath(ABOUT_PAGE_ADDL_INFO_CONFIG_PATH)) {

            String addlInfoPath = config.getString(ABOUT_PAGE_ADDL_INFO_CONFIG_PATH);
            Path addlInfoPathObj = Paths.get(addlInfoPath);

            try {

                byte[] addlInfoAsBytes = Files.readAllBytes(addlInfoPathObj);
                addlInfoAsHtml = new String(addlInfoAsBytes, "utf-8");

            } catch (IOException e) {
                throw new RuntimeException(e);
            }

        } else {

            addlInfoAsHtml = null;
        }

        return ok(views.html.about.render(PLAYFOP_URL, addlInfoAsHtml));
    }
}
